File: examples/ch10/comparison.js
/**
 * comparison.js - JavaScript fetch/axios示例
 * 功能：演示使用JavaScript fetch和axios进行HTTP请求
 * 对比：与CMD curl命令的对比
 * 
 * 注意：需要Node.js环境，axios需要安装: npm install axios
 */

// 使用fetch API（现代浏览器和Node.js 18+内置）
async function fetchExample() {
    console.log("=" .repeat(50));
    console.log("JavaScript fetch API示例");
    console.log("=" .repeat(50));
    console.log();
    
    // 1. GET请求
    console.log("[1] GET请求示例");
    console.log("-".repeat(30));
    
    try {
        // 基本GET请求
        const response = await fetch('https://httpbin.org/get');
        const data = await response.json();
        
        console.log(`状态码: ${response.status}`);
        console.log(`响应头: ${response.headers.get('Content-Type')}`);
        console.log(`响应数据: ${JSON.stringify(data).substring(0, 200)}...`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 2. 带参数的GET请求
    console.log("[2] 带参数的GET请求");
    console.log("-".repeat(30));
    
    const params = new URLSearchParams({
        name: 'test',
        value: '123'
    });
    
    try {
        const response = await fetch(`https://httpbin.org/get?${params}`);
        const data = await response.json();
        
        console.log(`完整URL: ${data.url}`);
        console.log(`状态码: ${response.status}`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 3. POST请求（JSON数据）
    console.log("[3] POST请求示例");
    console.log("-".repeat(30));
    
    const postData = {
        username: 'admin',
        password: 'secret123',
        email: 'admin@example.com'
    };
    
    try {
        const response = await fetch('https://httpbin.org/post', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(postData)
        });
        
        const data = await response.json();
        console.log(`状态码: ${response.status}`);
        console.log(`提交的数据: ${JSON.stringify(data.json)}`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 4. 设置请求头
    console.log("[4] 设置请求头示例");
    console.log("-".repeat(30));
    
    try {
        const response = await fetch('https://httpbin.org/get', {
            headers: {
                'User-Agent': 'JavaScript-Tutorial/1.0',
                'Accept': 'application/json',
                'Authorization': 'Bearer test_token_12345'
            }
        });
        
        const data = await response.json();
        console.log(`状态码: ${response.status}`);
        console.log(`请求头已设置: User-Agent, Accept, Authorization`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 5. 文件下载
    console.log("[5] 文件下载示例");
    console.log("-".repeat(30));
    
    try {
        const response = await fetch('https://httpbin.org/bytes/1024');
        const buffer = await response.arrayBuffer();
        
        console.log(`文件大小: ${buffer.byteLength} 字节`);
        console.log(`状态码: ${response.status}`);
        
        // 注意：在Node.js中可以使用fs模块保存文件
        // 在浏览器中可以使用Blob和URL.createObjectURL
    } catch (error) {
        console.error(`下载错误: ${error.message}`);
    }
    console.log();
    
    // 6. 错误处理
    console.log("[6] 错误处理示例");
    console.log("-".repeat(30));
    
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 3000); // 3秒超时
        
        const response = await fetch('https://nonexistent.example.com', {
            signal: controller.signal
        });
        
        clearTimeout(timeoutId);
    } catch (error) {
        if (error.name === 'AbortError') {
            console.log("请求超时");
        } else if (error.message.includes('Failed to fetch')) {
            console.log("连接错误: 无法连接到服务器");
        } else {
            console.log(`请求错误: ${error.message}`);
        }
    }
    console.log();
    
    // 7. 响应处理
    console.log("[7] 响应处理示例");
    console.log("-".repeat(30));
    
    try {
        const response = await fetch('https://httpbin.org/get');
        
        console.log(`状态码: ${response.status}`);
        console.log(`是否成功: ${response.ok}`);
        console.log(`内容类型: ${response.headers.get('Content-Type')}`);
        
        // 检查响应类型
        const contentType = response.headers.get('Content-Type');
        if (contentType && contentType.includes('application/json')) {
            const data = await response.json();
            console.log(`JSON数据: ${JSON.stringify(data)}`);
        }
    } catch (error) {
        console.error(`响应处理错误: ${error.message}`);
    }
    console.log();
    
    // 8. 批量请求
    console.log("[8] 批量请求示例");
    console.log("-".repeat(30));
    
    const urls = [
        'https://httpbin.org/get',
        'https://httpbin.org/ip',
        'https://httpbin.org/user-agent'
    ];
    
    try {
        const results = await Promise.all(
            urls.map(async url => {
                const response = await fetch(url);
                const data = await response.json();
                return {
                    url: url,
                    status: response.status,
                    data: data
                };
            })
        );
        
        results.forEach(result => {
            console.log(`URL: ${result.url}`);
            console.log(`  状态: ${result.status}`);
        });
    } catch (error) {
        console.error(`批量请求错误: ${error.message}`);
    }
    console.log();
}

// 使用axios库（需要安装：npm install axios）
async function axiosExample() {
    console.log("=" .repeat(50));
    console.log("JavaScript axios库示例");
    console.log("=" .repeat(50));
    console.log();
    
    // 动态导入axios（如果可用）
    let axios;
    try {
        axios = require('axios');
    } catch (error) {
        console.log("axios未安装，跳过axios示例");
        console.log("安装命令: npm install axios");
        console.log();
        return;
    }
    
    // 1. GET请求
    console.log("[1] axios GET请求");
    console.log("-".repeat(30));
    
    try {
        const response = await axios.get('https://httpbin.org/get');
        console.log(`状态码: ${response.status}`);
        console.log(`响应数据: ${JSON.stringify(response.data).substring(0, 200)}...`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 2. POST请求
    console.log("[2] axios POST请求");
    console.log("-".repeat(30));
    
    try {
        const response = await axios.post('https://httpbin.org/post', {
            username: 'admin',
            password: 'secret123'
        });
        
        console.log(`状态码: ${response.status}`);
        console.log(`提交的数据: ${JSON.stringify(response.data.json)}`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 3. 设置默认配置
    console.log("[3] axios默认配置");
    console.log("-".repeat(30));
    
    const api = axios.create({
        baseURL: 'https://httpbin.org',
        timeout: 5000,
        headers: {
            'User-Agent': 'JavaScript-Axios/1.0',
            'Accept': 'application/json'
        }
    });
    
    try {
        const response = await api.get('/get');
        console.log(`状态码: ${response.status}`);
        console.log(`请求头已设置: User-Agent, Accept`);
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
    
    // 4. 拦截器
    console.log("[4] axios拦截器");
    console.log("-".repeat(30));
    
    // 请求拦截器
    api.interceptors.request.use(config => {
        console.log(`发送请求: ${config.method.toUpperCase()} ${config.url}`);
        return config;
    });
    
    // 响应拦截器
    api.interceptors.response.use(response => {
        console.log(`收到响应: ${response.status} ${response.config.url}`);
        return response;
    });
    
    try {
        await api.get('/get');
    } catch (error) {
        console.error(`请求错误: ${error.message}`);
    }
    console.log();
}

// 与CMD curl对比
function compareWithCurl() {
    console.log("=" .repeat(50));
    console.log("与CMD curl对比");
    console.log("=" .repeat(50));
    console.log();
    
    console.log("GET请求:");
    console.log("  JavaScript: fetch('https://httpbin.org/get')");
    console.log("  CMD:        curl https://httpbin.org/get");
    console.log();
    
    console.log("POST请求:");
    console.log("  JavaScript: fetch(url, { method: 'POST', body: JSON.stringify(data) })");
    console.log("  CMD:        curl -X POST url -H \"Content-Type: application/json\" -d \"{}\"");
    console.log();
    
    console.log("设置头:");
    console.log("  JavaScript: fetch(url, { headers: { 'User-Agent': 'test' } })");
    console.log("  CMD:        curl -H \"User-Agent: test\" url");
    console.log();
    
    console.log("超时:");
    console.log("  JavaScript: fetch(url, { signal: AbortSignal.timeout(5000) })");
    console.log("  CMD:        curl --connect-timeout 5 url");
    console.log();
    
    console.log("错误处理:");
    console.log("  JavaScript: try { ... } catch (error) { ... }");
    console.log("  CMD:        if errorlevel 1 echo 错误");
    console.log();
}

// 主函数
async function main() {
    console.log("JavaScript HTTP请求示例");
    console.log("对比CMD curl命令");
    console.log();
    
    // 运行fetch示例
    await fetchExample();
    
    // 运行axios示例
    await axiosExample();
    
    // 显示对比
    compareWithCurl();
    
    console.log("=" .repeat(50));
    console.log("示例完成");
    console.log("=" .repeat(50));
}

// 运行主函数
main().catch(console.error);

File: examples/ch10/comparison.c
/**
 * comparison.c - C语言socket编程示例
 * 功能：演示使用C语言socket进行HTTP请求
 * 对比：与CMD curl命令的对比
 * 
 * 编译：gcc comparison.c -o comparison -lws2_32 (Windows)
 *       gcc comparison.c -o comparison (Linux/macOS)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
    #include <winsock2.h>
    #include <ws2tcpip.h>
    #pragma comment(lib, "ws2_32.lib")
#else
    #include <sys/socket.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
    #include <netdb.h>
    #include <unistd.h>
#endif

#define BUFFER_SIZE 4096
#define HTTP_PORT 80
#define HTTPS_PORT 443

// 初始化socket库（Windows需要）
int init_socket() {
#ifdef _WIN32
    WSADATA wsaData;
    return WSAStartup(MAKEWORD(2, 2), &wsaData);
#else
    return 0;
#endif
}

// 清理socket库
void cleanup_socket() {
#ifdef _WIN32
    WSACleanup();
#endif
}

// 关闭socket
void close_socket(int sock) {
#ifdef _WIN32
    closesocket(sock);
#else
    close(sock);
#endif
}

// 发送HTTP GET请求
int http_get(const char *host, const char *path, char *response, int response_size) {
    int sock;
    struct sockaddr_in server_addr;
    struct hostent *server;
    char request[BUFFER_SIZE];
    int bytes_received;
    
    // 创建socket
    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("创建socket失败");
        return -1;
    }
    
    // 解析主机名
    server = gethostbyname(host);
    if (server == NULL) {
        fprintf(stderr, "DNS解析失败: %s\n", host);
        close_socket(sock);
        return -1;
    }
    
    // 设置服务器地址
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(HTTP_PORT);
    memcpy(&server_addr.sin_addr.s_addr, server->h_addr, server->h_length);
    
    // 连接服务器
    if (connect(sock, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("连接服务器失败");
        close_socket(sock);
        return -1;
    }
    
    // 构建HTTP请求
    snprintf(request, sizeof(request),
        "GET %s HTTP/1.1\r\n"
        "Host: %s\r\n"
        "User-Agent: C-Tutorial/1.0\r\n"
        "Accept: text/html,application/json\r\n"
        "Connection: close\r\n"
        "\r\n",
        path, host);
    
    // 发送请求
    if (send(sock, request, strlen(request), 0) < 0) {
        perror("发送请求失败");
        close_socket(sock);
        return -1;
    }
    
    // 接收响应
    int total_received = 0;
    while (total_received < response_size - 1) {
        bytes_received = recv(sock, response + total_received, response_size - total_received - 1, 0);
        if (bytes_received <= 0) break;
        total_received += bytes_received;
    }
    response[total_received] = '\0';
    
    close_socket(sock);
    return total_received;
}

// 发送HTTP POST请求
int http_post(const char *host, const char *path, const char *data, char *response, int response_size) {
    int sock;
    struct sockaddr_in server_addr;
    struct hostent *server;
    char request[BUFFER_SIZE];
    int bytes_received;
    
    // 创建socket
    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("创建socket失败");
        return -1;
    }
    
    // 解析主机名
    server = gethostbyname(host);
    if (server == NULL) {
        fprintf(stderr, "DNS解析失败: %s\n", host);
        close_socket(sock);
        return -1;
    }
    
    // 设置服务器地址
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(HTTP_PORT);
    memcpy(&server_addr.sin_addr.s_addr, server->h_addr, server->h_length);
    
    // 连接服务器
    if (connect(sock, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("连接服务器失败");
        close_socket(sock);
        return -1;
    }
    
    // 构建HTTP POST请求
    snprintf(request, sizeof(request),
        "POST %s HTTP/1.1\r\n"
        "Host: %s\r\n"
        "User-Agent: C-Tutorial/1.0\r\n"
        "Content-Type: application/json\r\n"
        "Content-Length: %d\r\n"
        "Connection: close\r\n"
        "\r\n"
        "%s",
        path, host, (int)strlen(data), data);
    
    // 发送请求
    if (send(sock, request, strlen(request), 0) < 0) {
        perror("发送请求失败");
        close_socket(sock);
        return -1;
    }
    
    // 接收响应
    int total_received = 0;
    while (total_received < response_size - 1) {
        bytes_received = recv(sock, response + total_received, response_size - total_received - 1, 0);
        if (bytes_received <= 0) break;
        total_received += bytes_received;
    }
    response[total_received] = '\0';
    
    close_socket(sock);
    return total_received;
}

// 解析HTTP响应状态码
int parse_status_code(const char *response) {
    int status_code = 0;
    if (sscanf(response, "HTTP/%*s %d", &status_code) == 1) {
        return status_code;
    }
    return -1;
}

// 提取HTTP响应体
const char *extract_body(const char *response) {
    const char *body = strstr(response, "\r\n\r\n");
    if (body) {
        return body + 4; // 跳过"\r\n\r\n"
    }
    return NULL;
}

// 演示GET请求
void demo_get() {
    printf("[1] GET请求示例\n");
    printf("----------------------------------------\n");
    
    char response[BUFFER_SIZE];
    int received = http_get("httpbin.org", "/get", response, sizeof(response));
    
    if (received > 0) {
        int status = parse_status_code(response);
        printf("状态码: %d\n", status);
        printf("响应长度: %d 字节\n", received);
        
        const char *body = extract_body(response);
        if (body) {
            printf("响应体（前200字符）:\n%.200s\n", body);
        }
    } else {
        printf("请求失败\n");
    }
    printf("\n");
}

// 演示POST请求
void demo_post() {
    printf("[2] POST请求示例\n");
    printf("----------------------------------------\n");
    
    const char *json_data = "{\"username\":\"admin\",\"password\":\"secret123\"}";
    char response[BUFFER_SIZE];
    
    int received = http_post("httpbin.org", "/post", json_data, response, sizeof(response));
    
    if (received > 0) {
        int status = parse_status_code(response);
        printf("状态码: %d\n", status);
        printf("提交的数据: %s\n", json_data);
        
        const char *body = extract_body(response);
        if (body) {
            printf("响应体（前200字符）:\n%.200s\n", body);
        }
    } else {
        printf("请求失败\n");
    }
    printf("\n");
}

// 演示错误处理
void demo_error_handling() {
    printf("[3] 错误处理示例\n");
    printf("----------------------------------------\n");
    
    char response[BUFFER_SIZE];
    
    // 测试不存在的主机
    printf("测试不存在的主机...\n");
    int received = http_get("nonexistent.example.com", "/", response, sizeof(response));
    
    if (received < 0) {
        printf("错误: 无法连接到服务器\n");
    }
    printf("\n");
}

// 与CMD curl对比
void compare_with_curl() {
    printf("[4] 与CMD curl对比\n");
    printf("----------------------------------------\n");
    
    printf("GET请求:\n");
    printf("  C语言: http_get(\"httpbin.org\", \"/get\", response, sizeof(response))\n");
    printf("  CMD:   curl https://httpbin.org/get\n\n");
    
    printf("POST请求:\n");
    printf("  C语言: http_post(\"httpbin.org\", \"/post\", data, response, sizeof(response))\n");
    printf("  CMD:   curl -X POST https://httpbin.org/post -d \"data\"\n\n");
    
    printf("设置头:\n");
    printf("  C语言: 在请求字符串中添加\"Header: value\\r\\n\"\n");
    printf("  CMD:   curl -H \"Header: value\" url\n\n");
    
    printf("错误处理:\n");
    printf("  C语言: 检查返回值和socket错误\n");
    printf("  CMD:   if errorlevel 1 echo 错误\n\n");
}

int main() {
    printf("C语言socket编程示例\n");
    printf("对比CMD curl命令\n\n");
    
    // 初始化socket库
    if (init_socket() != 0) {
        fprintf(stderr, "初始化socket库失败\n");
        return 1;
    }
    
    printf("=" .repeat(50) "\n");
    printf("HTTP请求示例\n");
    printf("=" .repeat(50) "\n\n");
    
    // 运行示例
    demo_get();
    demo_post();
    demo_error_handling();
    compare_with_curl();
    
    printf("=" .repeat(50) "\n");
    printf("示例完成\n");
    printf("=" .repeat(50) "\n");
    
    // 清理socket库
    cleanup_socket();
    
    return 0;
}